### PR TITLE
Don't flow self contained command-line option to referenced projects

### PR DESCRIPTION
**Pull request by dsplaisted**
Fixes dotnet/sdk#21677

### Context

In .NET 6, if you run a command such as `dotnet build -r win-x64`, you will get the following error:

> C:\Program Files\dotnet\sdk\6.0.100-rtm.21474.29\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.Sdk.targets(1105,5): warning NETSDK1179: One of '--self-contained' or '--no-self-contained' options are required when '--runtime' is used. [C:\git\repro\ConsoleTest\ConsoleTest.csproj]

However, if you follow the suggestion and change it to `dotnet build -r win-x64 --self-contained`, then you will likely get an error such as the following if there are any referenced projects:

> C:\Program Files\dotnet\sdk\6.0.100-rc.2.21501.4\Sdks\Microsoft.NET.Sdk\targets\Microsoft.NET.RuntimeIdentifierInference.targets(150,5): error NETSDK1031: It is not supported to build or publish a self-contained application without specifying a RuntimeIdentifier. You must either specify a RuntimeIdentifier or set SelfContained to false. [C:\Users\brady\source\TodoApp\TodoApp.API\TodoApp.API.csproj]

This happens because

- Global properties (ie those specified on the command line) flow to referenced projects by default
- We normally stop the `RuntimeIdentifier` from flowing to referenced projects
- We don't currently stop the `SelfContained` property from flowing to referenced projects
- So the referenced project gets built with `SelfContained` set to true but no `RuntimeIdentifier`, which is an error

### Changes Made

If `--self-contained` is specified on the command-line, and the `RuntimeIdentifier` is blocked from flowing to referenced projects, then we also block the `SelfContained` property from flowing to referenced projects.

Note that this treats `--self-contained` differently from `/p:SelfContained=true`, as there is at least one project that relies on `/p:SelfContained=true` flowing across project references, and they set `RuntimeIdentifier` in their .props files if `SelfContained` is true.

### Testing

I will add automated tests to the SDK repo for this scenario, and test locally with those tests and the changes in this PR before merging.

### Notes

FYI @richlander @sfoslund @marcpopMSFT, I think we should take this for 17.0 GA

### CODE REVIEW

**Comment by BenVillalobos** [comment]
If `_CommandLineDefinedSelfContained` is the expected escape hatch,  why mark it with an underscore? 

I'm a little confused at why there's a distinction between --self-contained and /p:SelfContained=true, but I won't block on it.

**Comment by dsplaisted** [comment]
More background than you probably want to dig into is available here: https://github.com/dotnet/designs/blob/main/accepted/2021/architecture-targeting.md

We wouldn't expect someone to define `_CommandLineDefinedSelfContained` themselves, the CLI automatically sets it if you use the `--self-contained` or `--no-self-contained` options: https://github.com/dotnet/sdk/blob/f72371bb4e6b5d3394ec2c261cf8c43a6697b815/src/Cli/dotnet/CommonOptions.cs#L115-L125

So the escape hatch is basically to use `/p:SelfContained=true` instead of `--self-contained`.

@Forgind We shouldn't merge this yet until I've had a chance to test it.  I plan on writing automated tests in dotnet/sdk, but we won't be able to enable them in CI until the fix merges and flows.  But I should still run them manually first.

**Comment by dsplaisted** [comment]
I've done some more work and I think I have a pretty good idea of how we should solve this.  It involves changes to MSBuild (now included in this PR) as well as to the SDK, in dotnet/sdk#21986.

I'd like some feedback on whether that's the right direction.  It will be a breaking change (see the breaking change document in the SDK PR), so I think we *probably* would want to target these fixes for 6.0.200 rather than 6.0.100 GA.

**Comment by dsplaisted** [comment]
@rainersigwald I've retargeted this to vs17.1, and I think it (together with https://github.com/dotnet/sdk/pull/21986) is ready for review again

**Comment by rainersigwald** [comment]
Our primary 17.1 branch is `main` so I retargeted for you again. Reviewing.

**Comment by beatcracker** [comment]
I'm looking into this due to dotnet/sdk/issues/10566, and I've found that just checking for `_CommandLineDefinedRuntimeIdentifier` works for my simple case (no `rid` in the project, `self-contained` / `rid` are supplied via the commandline):

```xml
<IsRidAgnostic Condition=" '$(RuntimeIdentifier)' == '' and '$(RuntimeIdentifiers)' == '' and $(_CommandLineDefinedRuntimeIdentifier) == '' ">true</IsRidAgnostic>
```

Obviously, I've just skimmed the target source and I have no deep knowledge, but I hope this helps.

**Comment by Forgind** [comment]
@dsplaisted, are you going to be able to get around to this soon? Or should we close it for now or mark it as a draft?

**Comment by dsplaisted** [comment]
@Forgind Is there a way to convert a pull request back to a draft PR?

**Comment by dsplaisted** [comment]
@rainersigwald @dotnet/msbuild I've updated and rebased this PR, and I believe it is now ready for review again.  I switched back to using `IsRidAgnostic` instead of creating `AcceptsRuntimeIdentifier`, which should help make this more compatible.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1870]: When is '@(AnnotatedProjects)' != '%(Identity)'?

File: src/Tasks/Microsoft.Common.CrossTargeting.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CrossTargeting.targets:45]: Should GetTargetFrameworks depend on GetTargetFrameworksWithPlatformForSingleSingleTargetFramework explicitly?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1782]: It was previously RuntimeIdentifier$(_GlobalPropertiesToRemoveFromProjectReferences) (no semicolon)
That looks like a bug to me, but I just wanted to make sure that's intentional 🙂

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1952]: Should this be true and the next one false?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1945]: Does _TargetFrameworkInfo ever have more than just $(TargetFramework) included in it?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1870]: I think you have to do this when updating metadata this way, otherwise each update will apply to all the items, or something like that.  @rainersigwald probably knows the details.  Anyway my changes don't change that pattern here, they just change the other conditions.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (joeloff) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1798]: Do we need a `;` between `Configuration` and `$(_GlobalPropertiesToRemoveFromProjectReferences)`?

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (joeloff) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1964]: +1, it looks like it will set `IsRidAgnostic` to false when there are no runtime identifiers

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1798]: Looking at all usages of _GlobalPropertiesToRemoveFromProjectReferences in MSBuild, it seems that it is always expected to have a `;` in front of it. That said, I think adding a ; is better anyway.

File: src/Tasks/Microsoft.Common.CrossTargeting.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CrossTargeting.targets:45]: That's what it does for single-targeted projects: https://github.com/dotnet/msbuild/blob/5d102ae37163fc1f22610bb433c3ab9a2fff91f0/src/Tasks/Microsoft.Common.CurrentVersion.targets#L1889-L1890

For multi-targeted projects it has to do inner builds for each target framework, which the `GetTargetFrameworksWithPlatformFromInnerBuilds` target handles.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1782]: Good point.  It does seem like it might have been a bug before.  Now there may be extra semicolons, but that should be OK, right?  The MSBuild task should just ignore the extra ones I think.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1964]: Yes, thanks for catching this.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1957]: Yes, it does for multi-targeted projects, which will run this target for each target framework and combine the results.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1782]: Yes

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (Forgind) [inline src/Tasks/SetRidAgnosticValueForProjects.cs:29]: Why did you choose !false rather than true? Could it be null here? If so, this seems like wrong behavior.

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (Forgind) [inline src/Tasks/SetRidAgnosticValueForProjects.cs:31]: Should we have a Debug.Assert here or something?

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (Forgind) [inline src/Tasks/SetRidAgnosticValueForProjects.cs:36]: Are projects actually expected to have all of this information? I feel like this relies on projects carrying around a lot of (generally unimportant) baggage. Not saying that doesn't happen, but it's a bit surprising to me.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (Forgind) [inline src/Tasks/Microsoft.Common.CurrentVersion.targets:1865]: Why is this line necessary?

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (dsplaisted) [reply src/Tasks/SetRidAgnosticValueForProjects.cs:29]: If the value is not set I want this to be a no-op.  That seems the safest.

For projects using the .NET SDK, this should always be set.  For other types of projects it might not be.

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (dsplaisted) [reply src/Tasks/SetRidAgnosticValueForProjects.cs:31]: `Debug.Assert` gets compiled out in release builds, so that would be pretty unlikely to catch anything anyway, wouldn't it be?

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (dsplaisted) [reply src/Tasks/SetRidAgnosticValueForProjects.cs:36]: What information are you referring to?  These are part of the [Project Reference Protocol](https://github.com/dotnet/msbuild/blob/main/documentation/ProjectReference-Protocol.md#targets-required-to-be-referenceable), or in the case of `NearestTargetFramework`, set by the `GetReferenceNearestTargetFrameworkTask` task which runs just before this one.

File: src/Tasks/Microsoft.Common.CurrentVersion.targets
Comment (dsplaisted) [reply src/Tasks/Microsoft.Common.CurrentVersion.targets:1865]: It just clears it out so that it's not taking up memory anymore.  It's a pretty clunky pattern for when you want to have a task that updates item metadata.

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (Forgind) [reply src/Tasks/SetRidAgnosticValueForProjects.cs:36]: I wasn't aware of NearestTargetFramework, but that makes sense. I was also surprisedby IsRidAgnostic being an array of IsRidAgnostic values with one per target framework. I'd expected (and the Project Reference Protocol seems to say) that it's just a boolean. It seems like you're changing the type in this task from boolean[] to boolean?

File: src/Tasks/SetRidAgnosticValueForProjects.cs
Comment (Forgind) [reply src/Tasks/SetRidAgnosticValueForProjects.cs:31]: It is, but if something goes wrong, we'd probably try to debug with a debug build, so I thought it might be helpful. Not a big deal if you don't want to add it.

**Review by Forgind — APPROVED** [review]
I'm least confident about the IsRidAgnostic array, and I haven't had a chance to manually test this myself, but I don't know of anything broken in here, you said you wanted it in this week, and I have something in five minutes 🙂

**Pull request merged**
